torch/csrc/jit/script/init.cpp

```diff
@@ -915,19 +916,38 @@ void initJitScriptBindings(PyObject* module) {
         cu->register_type(classType);
         std::vector<ResolverPtr> rcbs;
         std::vector<Def> methodDefs;
-        for (const auto& def : classDef.body()) {
-          if (def.kind() != TK_DEF) {
-            throw ErrorReport(def.range())
+        std::vector<std::pair<Def, FunctionDefaults>> all_defaults;
+        for (const auto& def_stmt : classDef.body()) {
+          if (def_stmt.kind() != TK_DEF) {
+            throw ErrorReport(def_stmt.range())
                 << "Currently class bodies can only contain method "
                    "definitions. File an issue on Github if you want "
                    "something else!";
           }
-          methodDefs.emplace_back(Def(def));
+          Def def(def_stmt);
+          methodDefs.emplace_back(def_stmt);
           rcbs.push_back(
               pythonResolver(rcb, classDef.name().name(), classType));
+          auto defaults_for_def = name_to_defaults.find(def.name().name());
+          TORCH_INTERNAL_ASSERT(defaults_for_def != name_to_defaults.end());
+          all_defaults.push_back(std::make_pair<Def, FunctionDefaults>(
+              std::move(def), std::move(defaults_for_def->second)));
         }
         const auto self = SimpleSelf(classType);
-        cu->define(classname, methodDefs, rcbs, &self);
+        auto defined_fns = cu->define(classname, methodDefs, rcbs, &self);
```

Comment:
I don't think this works, because we're defining all of the methods on the class and then adding on the defaults. Any class method that calls another method with a default will fail.
```
        class X(object):
            def __init__(self, x=2, y=None):
                # type: (int, Optional[Tensor])
                self.x = x
                self.y = y

            def go(self, extra=2):
                # type: (int)
                return self.x, self.y, extra
  
            def will_fail(self):
                return self.go()
```